private_transactions/src/encryptor.rs

```diff
@@ -0,0 +1,318 @@
+		let decrypted_bytes = accounts.decrypt(requester, None, &ethcrypto::DEFAULT_MAC, &encrypted_bytes)?;
+		let decrypted_key = Public::from_slice(&decrypted_bytes);
+
+		// and now take x coordinate of Public as a key
+		let key: Bytes = (*decrypted_key)[..INIT_VEC_LEN].into();
+
+		// cache the key in the session and clear expired sessions
+		self.sessions.lock().insert(*contract_address, EncryptionSession{
+			key: key.clone(),
+			end_time: Instant::now() + Duration::from_millis(ENCRYPTION_SESSION_DURATION),
+		});
+		self.clean_expired_sessions();
+		Ok(key)
+	}
+
+	fn clean_expired_sessions(&self) {
+		let mut sessions = self.sessions.lock();
+		sessions.retain(|_, session| session.end_time < Instant::now());
+	}
+
+	fn obtained_key(&self, contract_address: &Address) -> Option<Bytes> {
+		let mut sessions = self.sessions.lock();
+		let stored_session = sessions.entry(*contract_address);
+		match stored_session {
+			Entry::Occupied(session) => {
+				if Instant::now() > session.get().end_time {
+					session.remove_entry();
+					None
+				} else {
+					Some(session.get().key.clone())
+				}
+			}
+			Entry::Vacant(_) => None,
+		}
+	}
+
+	/// Try to unlock account using stored passwords
+	fn unlock_account(&self, account: &Address, accounts: Arc<AccountProvider>) -> Result<bool, PrivateTransactionError> {
+		let passwords = self.config.passwords.clone();
+		for password in passwords {
+			if let Ok(()) = accounts.unlock_account_temporarily(account.clone(), password) {
+				return Ok(true);
+			}
+		}
+		Ok(false)
+	}
+
+	fn sign_contract_address(&self, contract_address: &Address, accounts: Arc<AccountProvider>) -> Result<Signature, PrivateTransactionError> {
+		// key id in SS is H256 && we have H160 here => expand with assitional zeros
+		let contract_address_extended: H256 = contract_address.into();
+		let key_server_account = self.config.key_server_account.ok_or_else(|| PrivateTransactionError::KeyServerAccountNotSet)?;
+		if let Ok(true) = self.unlock_account(&key_server_account, accounts.clone()) {
+			Ok(accounts.sign(key_server_account.clone(), None, H256::from_slice(&contract_address_extended))?)
+		} else {
+			trace!("Cannot unlock account");
+			Err(PrivateTransactionError::Encrypt("Cannot unlock account".into()))
+		}
+	}
+}
+
+impl Encryptor for SecretStoreEncryptor {
+	fn encrypt(
+		&self,
+		contract_address: &Address,
+		accounts: Arc<AccountProvider>,
+		initialisation_vector: &H128,
+		plain_data: &[u8]
+	) -> Result<Bytes, PrivateTransactionError> {
+		// retrieve the key, try to generate it if it doesn't exist yet
+		let key = match self.retrieve_key("", false, contract_address, accounts.clone()) {
+			Ok(key) => Ok(key),
+			Err(PrivateTransactionError::EncryptionKeyNotFound(_)) => {
+				trace!("Key for account wasnt found in sstore. Creating. Address: {:?}", contract_address);
+				self.retrieve_key(&format!("/{}", self.config.threshold), true, contract_address, accounts.clone())
+			}
+			Err(err) => Err(err),
```

Comment:
```rust
if let Err(PrivateTransactionError::EncryptionKeyNotFound(_)) = key {
  // do what you do
}
```

Reply:
It's not equivalent to the current code, because the error needs to be propagated